### PR TITLE
Introduce `recordVersion` for records in protocol

### PR DESCRIPTION
**Pull request by korthout**
> **Warning**
> This pull request touches code maintained by @camunda/zeebe-distributed-platform (`stream-processor`, and `backup` scopes), @camunda/zeebe-process-automation (`engine`, `exporters` scopes), and code that is shared between the two (`protocol`, `protocol-impl`, `test-util`, and commits without a scope). It should be reviewed by both teams.

## Description

This introduces a new property to records: `recordVersion`. This is an integer starting at `1` that can be incremented when we want to write a new version of a record. It also allows writing specific versions of records in case that is needed (required for #11661). The property is stored as record metadata.

Record versions are necessary to implement:
- #12764 

An attempt was made to use the existing `brokerVersion` to implement it in #11970 but this is hard to maintain (see https://github.com/camunda/zeebe/pull/11970#issuecomment-1544447868). It is never really clear what the next broker version will be while working on it. So it's hard to refer to it. Instead, it's much easier to have a counter that can be incremented for new versions.

The record version property is optional, and when missing the value `1` will be assumed. So all records that were written before this property existed, will be assumed to have version `1`.

## Related issues

part of #12764 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [x] [Operate](https://github.com/camunda/operate/issues)
- [x] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [x] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Review by npepinpe — APPROVED** [review]
:+1: 

Curious to see it in use!

Only had one important comment, but don't think we need a second review, so pre-approved!

File: protocol/src/main/resources/protocol.xml
Comment (remcowesterhoud) [inline protocol/src/main/resources/protocol.xml:143]: 🔧 We should move this up a few lines. It's now under `<!-- populated when RecordType is COMMAND_REJECTION -->`. But this is also filled for records that aren't command rejections. I'd argue the line above should also be moved up tbh.

File: protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/RecordMetadata.java
Comment (remcowesterhoud) [inline protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/RecordMetadata.java:94]: ❓ I'm a bit confused how this works. The default of the `recordVersion` is 1. So this means it is always filled for new records. For older records where we don't have a `recordVersion` what does this decoder return?

```java
    public int recordVersion()
    {
        if (parentMessage.actingVersion < 3)
        {
            return 65535;
        }

        return (buffer.getShort(offset + 30, java.nio.ByteOrder.LITTLE_ENDIAN) & 0xFFFF);
    }
```

It seems it can return `65535` which is higher than 0 as well, but I'm not sure what this if check means.

File: protocol-impl/src/test/java/io/camunda/zeebe/protocol/impl/JsonSerializableToJsonTest.java
Comment (remcowesterhoud) [inline protocol-impl/src/test/java/io/camunda/zeebe/protocol/impl/JsonSerializableToJsonTest.java:171]: ❌ We should add a testcase that verifies if no `recordVersion` is set, it should resolve to 1.

File: exporters/elasticsearch-exporter/src/main/resources/zeebe-record-template.json
Comment (remcowesterhoud) [inline exporters/elasticsearch-exporter/src/main/resources/zeebe-record-template.json:47]: 👍 Nice, I wouldn't have thought of these 🚀

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
Nice job @korthout 🏆

I've got a small suggestion and 1 blocker for a missing test case. All small stuff 😄

File: protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/RecordMetadata.java
Comment (korthout) [reply protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/RecordMetadata.java:94]: I'm not sure about this, I've asked the team [on slack](https://camunda.slack.com/archives/CSQ2E3BT4/p1685971972989149). I think we can ignore it, as our UpdateTests should cover processing (and applying) of records written on old versions of Zeebe.

Comment (korthout) [reply protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/RecordMetadata.java:94]: So, 65535 is the nullValue. See https://www.fixtrading.org/standards/sbe-online/#integer-encoding

This means that @remcowesterhoud found a bug in my code 🎉

**Comment by korthout** [comment]
Since I've made some changes, please have another look :bow

**Review by remcowesterhoud — APPROVED** [review]
🎉 LGTM! Thanks for the changes 🙇

**Comment by remcowesterhoud** [comment]
bors merge

**Comment by korthout** [comment]
Oh @npepinpe is out today, but he also said:
>but don't think we need a second review, so pre-approved!

So let's move forward with this. Any problems discovered later can also be handled in a separate pull request.

bors merge

File: protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/RecordMetadata.java
Comment (korthout) [reply protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/RecordMetadata.java]: Oops, it seems that using 65535 instead of 0 actually caused a bug 😅 See b981024.

**Pull request merged**
